rapidhash.h

```diff
@@ -257,82 +372,58 @@ RAPIDHASH_INLINE_CONSTEXPR uint64_t rapidhash_internal(const void *key, size_t l
-      see2 = rapid_mix(rapid_read64(p + 32) ^ secret[2], rapid_read64(p + 40) ^ see2);
-      p += 48;
-      i -= 48;
-      if (_likely_(i >= 48)) {
-        seed = rapid_mix(rapid_read64(p) ^ secret[0], rapid_read64(p + 8) ^ seed);
-        see1 = rapid_mix(rapid_read64(p + 16) ^ secret[1], rapid_read64(p + 24) ^ see1);
-        see2 = rapid_mix(rapid_read64(p + 32) ^ secret[2], rapid_read64(p + 40) ^ see2);
-        p += 48;
-        i -= 48;
-      }
-    }
-    see3 ^= see4;
-    see5 ^= see6;
-    seed ^= see1;
-    see3 ^= see2;
-    seed ^= see5;
-    seed ^= see3;
-    if ((i > 16)) {
-      seed = rapid_mix(rapid_read64(p) ^ secret[2], rapid_read64(p + 8) ^ seed);
-      if (_unlikely_(i > 32))
-        seed = rapid_mix(rapid_read64(p + 16) ^ secret[2], rapid_read64(p + 24) ^ seed);
-    }
-    a=rapid_read64(p+i-16);  b=rapid_read64(p+i-8);
-  } else {
+  } else if (_likely_(len <= 64)) {
+    // input len [17, 64]
     seed = rapid_mix(rapid_read64(p) ^ secret[0], rapid_read64(p + 8) ^ seed);
-    if (len > 32) {
+    if ((len > 32)) {
       seed = rapid_mix(rapid_read64(p + 16) ^ secret[1], rapid_read64(p + 24) ^ seed);
       if (len > 48) {
         seed = rapid_mix(rapid_read64(p + 32) ^ secret[0], rapid_read64(p + 40) ^ seed);
       }
     }
     a=rapid_read64(p+len-16);  b=rapid_read64(p+len-8);
+  } else if (len <= 288) {
+    // input len [65, 288] — using the original rapidhash V1 logic
+    size_t i = len;
+    uint64_t see1=seed, see2=seed;
+    while(_likely_(i>=96)){
+      seed=rapid_mix(rapid_read64(p)^secret[0],rapid_read64(p+8)^seed);
+      see1=rapid_mix(rapid_read64(p+16)^secret[1],rapid_read64(p+24)^see1);
+      see2=rapid_mix(rapid_read64(p+32)^secret[2],rapid_read64(p+40)^see2);
+      seed=rapid_mix(rapid_read64(p+48)^secret[0],rapid_read64(p+56)^seed);
+      see1=rapid_mix(rapid_read64(p+64)^secret[1],rapid_read64(p+72)^see1);
+      see2=rapid_mix(rapid_read64(p+80)^secret[2],rapid_read64(p+88)^see2);
+      p+=96; i-=96;
+    }
+    if(_unlikely_(i>=48)){
+      seed=rapid_mix(rapid_read64(p)^secret[0],rapid_read64(p+8)^seed);
+      see1=rapid_mix(rapid_read64(p+16)^secret[1],rapid_read64(p+24)^see1);
+      see2=rapid_mix(rapid_read64(p+32)^secret[2],rapid_read64(p+40)^see2);
+      p+=48; i-=48;
+    }
+    seed^=see1^see2;
+    if(i>16){
+      seed=rapid_mix(rapid_read64(p)^secret[2],rapid_read64(p+8)^seed^secret[1]);
+      if(i>32)
+        seed=rapid_mix(rapid_read64(p+16)^secret[2],rapid_read64(p+24)^seed);
+    }
+    a=rapid_read64(p+i-16);  b=rapid_read64(p+i-8);
+  } else [[unlikely]] {
+    // input len > 288
+    return rapidhash_internal_cold(p, len, seed, secret);
   }
+
   a ^= secret[1];
   b ^= seed;
-  rapid_mum(&a, &b);
-  return rapid_mix(a ^ secret[7] ^ len, b ^ secret[1]);
+
+  #ifdef RAPIDHASH_AVALANCHE
+    rapid_mum(&a, &b);
+    return rapid_mix(a ^ secret[7] ^ len, b ^ secret[1]);
+  #else
+    return rapid_mix(a ^ len, b);
+  #endif

```

Comment:
When not avalanching, we still XOR the length. Perhaps it's unnecessary as it's already included in the seed?

Reply:
This is where we `^ seed` instead of `^ len` on the understanding that the seed has already used the length. In rust, this final avalanching step is done in a separate method that isn't always inlined alongside the rest of the code, and so using the `seed` removes the need to store and update a long-lived length variable.